fix #44735, raise exception on `Vararg`-arguments in `subtype` rather than abort

aviatesk opened the pull request.

aviatesk commented:
@vtjnash does this seem sensible to you?

vtjnash commented:
I am not precisely sure how Vararg works now, but seems probably safe. Seems like something that should be handled internally to type-intersection though, where other conditions are checked?

aviatesk commented:
yeah, if `jl_type_intersection` does really only expect `Type`-objects, I think it might be better to change the signature to `typeintersect(@nospecialize(a::Type), @nospecialize(b::Type))`, and remove the vararg assertion from `jl_type_intersection`?

aviatesk commented:
ok, now the exception would be raised within subtype.c.

Keno commented:
IIRC @JeffBezanson felt pretty strongly that passing non-type object here is a bug and should assert, but maybe I'm misremembering. Regardless, we should check with him when he's back from vacation.

JeffBezanson commented:
Right, I don't think subtype/intersection should need to throw any errors. Currently there is only one (`circular type parameter constraint`), and it's basically a misplaced type-definition-time check.

I think we wanted this assert after the `TypeofVararg` change to make sure we caught all the cases that treated it like a type. Now that it's been working for a while, I think we could just remove the assert and let it fall through to `egal` like all other random values. If we later want to add checks that the arguments to `typeintersect` at the julia level are `Type`s, that might be ok too.

aviatesk committed:
remove `isvarargtype` assertions from `subtype` and `intersect`

fix #44735

aviatesk commented:
Thanks @JeffBezanson for your comment. I removed the `jl_is_vararg` assertions from subtype.c, and also added Julia level assertions that checks if input arguments are `Type`s.
I will try the `inference` benchmark: I will remove the Julia level error checks if benchmark result shows it can lead to some performance problem.

@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1089821608) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d39f6b_vs_2c8c0a4/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1089838979) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d39f6b_vs_2c8c0a4/report.md).

JeffBezanson commented on base/reflection.jl:
```diff
@@ -675,12 +675,17 @@ struct type with no fields.
+function typeintersect(@nospecialize(a), @nospecialize(b))
+    @_pure_meta
+    isa(a, Type) || throw(TypeError(:typeintersect, Type, typeof(a)))
+    isa(b, Type) || throw(TypeError(:typeintersect, Type, typeof(b)))
```
```suggestion
    isa(b, Type) || throw(TypeError(:typeintersect, Type, b))
```

vtjnash commented on base/reflection.jl:
```diff
@@ -675,12 +675,17 @@ struct type with no fields.
-typeintersect(@nospecialize(a), @nospecialize(b)) = (@_pure_meta; ccall(:jl_type_intersection, Any, (Any, Any), a, b))
+function typeintersect(@nospecialize(a), @nospecialize(b))
+    @_pure_meta
+    isa(a, Type) || throw(TypeError(:typeintersect, Type, typeof(a)))
```
Why isn't this written `a::Type`?

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1091593989) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4ed1360_vs_badad9d/report.md).

aviatesk commented:
Seems good to go?
